### ISSUE TITLE
indeterminate progress (aka spinner)

### ISSUE THREAD

**Issue by jpscharf** [thephpleague/climate#70]
Will a spinner or indeterminate progress ever be added?

**Comment by joetannenbaum** [comment]
I have definitely thought about it before, yes. Might be the next thing in the works....

**Comment by masnug** [comment]
how about this one, https://github.com/smasty/CLImate.php

**Comment by ValeriyLiquid** [comment]
+1 for this feature

**Comment by LostKobrakai** [comment]
Wasn't to hard to adapt the progress object to this.
```php
<?php

class Spinner extends DynamicTerminalObject
{
    /**
     * The sequence of chars to be spinned
     *
     * @var string $sequence
     */
    protected $sequence = '-\|/';

    /**
     * The current item of the sequence
     *
     * @var integer $current
     */
    protected $current = 0;

    /**
     * Flag indicating whether we are writing the bar for the first time
     *
     * @var boolean $first_line
     */
    protected $first_line = true;

    /**
     * Current label
     *
     * @var string $label
     */
    protected $label;

    /**
     * If they pass in a sequence, set the sequence
     *
     * @param integer $sequence
     */
    public function __construct($sequence = null, $label = null)
    {
        if ($sequence !== null) {
            $this->sequence($sequence);
        }
        if ($label !== null) {
             $this->label = $label;
        }
    }

    /**
     * Set the sequence property
     *
     * @param  string $sequence
     *
     * @return Progress
     */
    public function sequence($sequence)
    {
        $this->sequence = $sequence;

        return $this;
    }

    /**
     * Re-writes the spinner
     *
     * @param integer $current
     * @param mixed   $label
     * @throws \Exception
     */
    public function current($label = null)
    {
        $label = $label ?: $this->label;

        if (strlen($this->sequence) == 0) {
            // Avoid dividing by 0
            throw new \Exception('The progress total must be greater than zero.');
        }
        $current = ($this->current + 1) % strlen($this->sequence);

        $this->drawSpinner($current, $label);

        $this->current = $current;
        $this->label   = $label;
    }

    /**
     * Draw the spinner
     *
     * @param string $current
     * @param string $label
     */
    protected function drawSpinner($current, $label)
    {
        if ($this->first_line) {
            // Drop down a line, we are about to
            // re-write this line for the progress bar
            $this->output->write('');
            $this->first_line = false;
        }

        // Move the cursor up one line and clear it to the end
        $spinner  = $this->util->cursor->up(1);
        $spinner .= $this->util->cursor->startOfCurrentLine();
        $spinner .= $this->util->cursor->deleteCurrentLine();
        $spinner .= $this->getSpinner($current, $label);

        $this->output->write($this->parser->apply($spinner));
    }

    /**
     * Get the spinner string
     *
     * @param integer $current
     * @param string $label
     *
     * @return string
     */
    protected function getSpinner($current, $label)
    {
        $percentage = $current / strlen($this->sequence);

        $char = substr($this->sequence, $current, 1);

        if ($label) $label = " $label";

        return trim("{$char}{$label}");
    }
}
```

```php
<?php

$climate = new League\CLImate\CLImate;
$climate->extend('Spinner');
$spinner = $climate->spinner('-\|/', 'Searching for file.');
foreach (range(0, 10000) as $i) {
	$spinner->current();
	usleep(500);
}
```

**Comment by Oxicode** [comment]
+1 for this feature

**Comment by duncan3dc** [comment]
Hi everybody, can you try out the test version below and let me know your thoughts:
`composer require league/climate:dev-spinner`

Here's an example:
```php
$spinner = $climate->spinner('Working');
foreach (range(0, 10000) as $i) {
    $spinner->advance();
    usleep(1000);
}
```

And a version of the previous example from this thread:
```php
$spinner = $climate->spinner('Searching for file.');
$spinner->characters('-', '\\', '|', '/');
$spinner->timeLimit(0);
foreach (range(0, 10000) as $i) {
    $spinner->advance();
    usleep(1000);
}
```

**Comment by duncan3dc** [comment]
This has been merged into master now and will be part of the next release

**Comment by duncan3dc** [comment]
This has been released in version [3.5.0](https://github.com/thephpleague/climate/releases/tag/3.5.0)
